### PR TITLE
Fix bug in Planner

### PR DESCRIPTION
**Pull request by kasiafi**
Produced plans are incorrect in the case when:
there are multiple references to a correlated symbol in a correlated subquery
and different occurrences of the symbol require different coercions.
In the resulting plan all occurrences of the symbol are coerced in the way that was appropriate for the first occurrence.
This behavior is caused by `SubqueryPlanner` who collects correlated symbols using equality comparison, and as a result equates different occurrences of the symbols.
Solution: identify them by `NodeRef` instead of `Expression`.

### CODE REVIEW

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java [hunk 1]
```diff
@@ -3220,6 +3220,43 @@ public void testScalarSubquery()
         assertQuery("SELECT * FROM (VALUES 1.0) t(a) WHERE 1=(SELECT count(*) WHERE 1 = a)", "SELECT 1.0");
     }
 
+    @Test
+    public void testMultipleOccurrencesOfCorrelatedSymbol()
```
Comment (kokosing) [inline presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: can you please move these tests to `TestSubqueries`. ATQ tests are run per most of the connectors, while tests are focused only on planner.

File: presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/*
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java]: `Fix bug in Planner` -> `Fix coercion handling in SubqueryPlanner`?

File: presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    public static Expression replaceExpression(Expression expression, Map<NodeRef<Expression>, Expression> mappings)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new IdentitySensitiveExpressionNodeInliner(mappings), expression);
+    }
+
+    private final Map<NodeRef<Expression>, Expression> mappings;
+
+    public IdentitySensitiveExpressionNodeInliner(Map<NodeRef<Expression>, Expression> mappings)
+    {
+        this.mappings = mappings;
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java]: rnn and ImmutableMap.copyOf?

File: presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+package io.prestosql.sql.planner;
+
+import io.prestosql.sql.tree.Expression;
+import io.prestosql.sql.tree.ExpressionRewriter;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+import io.prestosql.sql.tree.NodeRef;
+
+import java.util.Map;
+
+public class IdentitySensitiveExpressionNodeInliner
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java]: `Identity` has special meaning in Presto. Maybe `NodeRefAwareExpressionNodeInliner`?

File: presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (kokosing) [inline presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java]: > This behavior is caused by `SubqueryPlanner` who collects correlated symbols
using equality comparison, and as a result equates different occurrences
of the symbols.

`NodeRef` was introduced later than `SubqueryPlanner` IIRC.

File: presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java (hunk 4 above)
Comment (kasiafi) [reply presto-main/src/main/java/io/prestosql/sql/planner/IdentitySensitiveExpressionNodeInliner.java]: Renamed to `ReferenceAwareExpressionNodeInliner`. I believe that class names shouldn't refer to technicalities.

File: presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java (hunk 1 above)
Comment (kasiafi) [reply presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: I am aware of that. I will move them as soon as tests with tables can be used in `TestSubqueries`.

Comment (kokosing) [reply presto-tests/src/main/java/io/prestosql/tests/AbstractTestQueries.java]: I was thinking of or replacing `nation` and `region` with `VALUES`.

**Comment by kokosing** [comment]
Merged, thanks!

**Pull request merged**